### PR TITLE
Underline styles and colors - core part

### PR DESCRIPTION
**Pull request by jerch**
PR to support additional underline styles and colors in `AttributeData`. Implements core parts of #1145.

Note that this PR extends `AttributeData` in a generic way, which can be use to store other data later on (like #1134).

Changes to the internal interface:
```TS
interface IAttributeData {
  ...
  // whether the attr object has any extended attrs defined
  hasExtendedAttrs(): number;
  // update internal HAS_EXTENDED flag
  updateExtended(): void;
  // transparently either returns FG color or underline color if set
  getUnderlineColor(): number;
  // color mode accessors
  getUnderlineColorMode(): number;
  isUnderlineColorRGB(): boolean;
  isUnderlineColorPalette(): boolean;
  isUnderlineColorDefault(): boolean;
  // actual underline
  getUnderlineStyle(): UnderlineStyle;
}
```

The color and the color mode accessors transparently map to the FG variants, if no independent color was set or underline is unset. Thus its important to test beforehand for `isUnderline`.

TODO:
- [x] handle extended SGR 4 sequences
- [x] handle SGR 58 and SGR 59 sequences
- [x] extend `AttributeData` to expose extended attributes
- [x] extend `BufferLine` to hold and export extended attributes
- [x] test cases for all

### CODE REVIEW

File: src/common/Types.d.ts [hunk 1]
```diff
@@ -98,6 +106,16 @@ export interface IAttributeData {
...
+
+  // extended attrs
+  hasExtendedAttrs(): number;
+  updateExtended(): void;
+  getUnderlineColor(): number;
+  getUnderlineColorMode(): number;
+  isUnderlineColorRGB(): boolean;
+  isUnderlineColorPalette(): boolean;
+  isUnderlineColorDefault(): boolean;
+  getUnderlineStyle(): number;
```
Comment (Tyriar) [inline src/common/Types.d.ts]: Shouldn't these be encapsulated within `IExtendedAttrs`?

File: src/common/Types.d.ts [hunk 2]
```diff
@@ -68,10 +68,18 @@ export interface ICharset {
...
+  underlineColor: number;
+  clone(): IExtendedAttrs;
+  isEmpty(): boolean;
+}
+
 /** Attribute data */
 export interface IAttributeData {
   fg: number;
   bg: number;
+  extended: IExtendedAttrs;
```
[the remark this answers is not in the document]
Comment (jerch) [reply src/common/Types.d.ts]: I had it as optional before, which lead to toxic deopts in the input chain, guess due to shape shifting of the attr object (tried with both `undefined` and `null`). So now this is a 1:1 object relation. Do you issues in that?

Offtopic: Btw I found the reduced timeline resolution issue by playing with this.

File: src/common/Types.d.ts (hunk 1 above)
Comment (jerch) [reply src/common/Types.d.ts]: They refer to data on `AttributeData`, which `IExtendedAttrs` is not aware of. (Currently `IExtendedAttrs` extends `AttributeData`, not the way around).

File: src/InputHandler.ts [hunk 3]
```diff
@@ -2026,6 +2026,21 @@ export class InputHandler extends Disposable implements IInputHandler {
     }
   }
 
+  /**
+   * Helper to write color information packed with color mode.
+   */
+  private _updateColor(color: number, mode: number, c1: number, c2: number, c3: number): number {
```
Comment (Tyriar) [inline src/InputHandler.ts:2032]: Little more specific?

```suggestion
  private _updateAttrColor(color: number, mode: number, c1: number, c2: number, c3: number): number {
```

File: src/InputHandler.ts [hunk 4]
```diff
@@ -2115,13 +2152,13 @@ export class InputHandler extends Disposable implements IInputHandler {
...
    * | 3         | Italic.                                                  | #Y      |
-   * | 4         | Underlined. (no support for newer underline styles)      | #Y      |
+   * | 4         | Underlined (see below for style support).                | #Y      |
    * | 5         | Slowly blinking.                                         | #N      |
    * | 6         | Rapidly blinking.                                        | #N      |
    * | 7         | Inverse. Flips foreground and background color.          | #Y      |
    * | 8         | Invisible (hidden).                                      | #Y      |
    * | 9         | Crossed-out characters.                                  | #N      |
-   * | 21        | Doubly  underlined.                                      | #N      |
+   * | 21        | Doubly  underlined.                                      | #P[Currently outputs a single underline.] |
```
Comment (Tyriar) [inline src/InputHandler.ts:2161]: 😱 

```suggestion
   * | 21        | Doubly underlined.                                       | #P[Currently outputs a single underline.] |
```

File: src/common/Types.d.ts (hunk 1 above)
Comment (Tyriar) [inline src/common/Types.d.ts:118]: Just to clarify, all underline styles will still result in the `isUnderline` public API being non-zero?

File: src/common/buffer/CellData.ts [hunk 5]
```diff
@@ -19,10 +19,11 @@ export class CellData extends AttributeData implements ICellData {
...
   }
   /** Primitives from terminal buffer. */
-  public content: number = 0;
-  public fg: number = 0;
-  public bg: number = 0;
-  public combinedData: string = '';
+  public content = 0;
+  public fg = 0;
+  public bg = 0;
+  public extended: IExtendedAttrs = new ExtendedAttrs();
```
Comment (Tyriar) [inline src/common/buffer/CellData.ts:25]: So the approximate memory increase per cell will be 1 pointer for this, and then the standard `ExtendedAttrs` object will get shared between all cells?

**Comment by jerch** [comment]
> Looks good after the minor issues are resolved +1, next PR would be adding rendering?

Yepp. But thats non trivial, as you already might have guessed from the comments in #1145. Gonna fix the issues above in the next couple of days, but prolly wont make it to the renderer parts :disappointed:

**Comment by Tyriar** [comment]
@jerch no worries, we should still merge to avoid conflicts piling up

File: src/common/Types.d.ts (hunk 1 above)
Comment (jerch) [reply src/common/Types.d.ts:134]: The idea here for renderer side is to first test for `isUnderline`, if that evals to `true`, you'd have to request the style and the color to finally draw it. This way it is compatible to our current code, which only relies on a `isUnderline` test and always draws a single line.

The color methods cannot be used to determine whether underline is in place, as they would always report a color (fg color). The style method can be used for that, yes.

File: src/common/buffer/CellData.ts (hunk 5 above)
Comment (jerch) [reply src/common/buffer/CellData.ts:25]: But only for cells, that actually have extended attribs set. All other (thus most cells) dont use more memory in the buffer. To get this working I used another bit in `bg` as indicator whether a cell holds extended attrs (`HAS_EXTENDED` flag).

**Pull request merged**